File: coldatomlab/web/camera3d.js
"use strict";
// Frozen-state camera, double-precision browser arithmetic; independent of GPE evolution.
window.AbsorptionCamera3D = {
  defaults: {
    axis: "z",
    binning: 1,
    fwhm_um: 0,
    saturation: 1,
    exposure_us: 5,
    efficiency: 0.8,
    read_noise: 1,
    noise: false,
    seed: 17,
    roi_um: 12,
    strip_um: 4,
  },
  sigma: (3 * 0.780241209 ** 2) / (2 * Math.PI),
  validate(c, n) {
    if (!["x", "y", "z"].includes(c.axis))
      throw Error("Choose an imaging axis.");
    if (
      ![1, 2, 4, 8, 16].includes(c.binning) ||
      n % c.binning ||
      n / c.binning < 8
    )
      throw Error(
        "Choose a pixel size with at least eight camera pixels per side.",
      );
    if (
      typeof c.noise !== "boolean" ||
      !Number.isInteger(c.seed) ||
      c.seed < 0 ||
      c.seed > 4294967295
    )
      throw Error("Noise needs a boolean flag and a 32-bit integer seed.");
    for (const [key, lo, hi] of [
      ["fwhm_um", 0, 20],
      ["saturation", 0.001, 10],
      ["exposure_us", 0.1, 100],
      ["efficiency", 0.05, 1],
      ["read_noise", 0, 20],
      ["roi_um", 1, 100],
      ["strip_um", 0.2, 40],
    ])
      if (
        typeof c[key] !== "number" ||
        !Number.isFinite(c[key]) ||
        c[key] < lo ||
        c[key] > hi
      )
        throw Error(`${key} must be between ${lo} and ${hi}.`);
  },
  rng(seed) {
    let state = seed >>> 0;
    const uniform = () => {
      state = (Math.imul(1664525, state) + 1013904223) >>> 0;
      return (state + 0.5) / 4294967296;
    };
    const normal = () =>
      Math.sqrt(-2 * Math.log(uniform())) * Math.cos(2 * Math.PI * uniform());
    const logFact = (k) => {
      if (k < 256) {
        let sum = 0;
        for (let j = 2; j <= k; j++) sum += Math.log(j);
        return sum;
      }
      return (
        (k + 0.5) * Math.log(k) -
        k +
        0.5 * Math.log(2 * Math.PI) +
        1 / (12 * k) -
        1 / (360 * k ** 3) +
        1 / (1260 * k ** 5)
      );
    };
    const poisson = (mu) => {
      if (mu < 30) {
        let k = 0,
          p = 1,
          bound = Math.exp(-mu);
        do {
          k++;
          p *= uniform();
        } while (p > bound);
        return k - 1;
      }
      const b = 0.931 + 2.53 * Math.sqrt(mu),
        a = -0.059 + 0.02483 * b,
        inv = 1.1239 + 1.1328 / (b - 3.4),
        vr = 0.9277 - 3.6224 / (b - 2);
      for (;;) {
        const u = uniform() - 0.5,
          v = uniform(),
          us = 0.5 - Math.abs(u),
          k = Math.floor(((2 * a) / us + b) * u + mu + 0.43);
        if (us >= 0.07 && v <= vr) return k;
        if (k < 0 || (us < 0.013 && v > us)) continue;
        if (
          Math.log((v * inv) / (a / (us * us) + b)) <=
          -mu + k * Math.log(mu) - logFact(k)
        )
          return k;
      }
    };
    return { uniform, normal, poisson };
  },
  transmission(od, s) {
    let lo = Math.max(0, od - s),
      hi = od;
    for (let j = 0; j < 52; j++) {
      const u = (lo + hi) / 2;
      if (u + s * -Math.expm1(-u) < od) lo = u;
      else hi = u;
    }
    return Math.exp(-(lo + hi) / 2);
  },
  blur(input, sigma) {
    if (sigma < 1e-8) return input.map((r) => r.slice());
    const n = input.length,
      r = Math.ceil(6 * sigma),
      k = Array.from({ length: 2 * r + 1 }, (_, i) =>
        Math.exp(-0.5 * ((i - r) / sigma) ** 2),
      ),
      sum = k.reduce((a, b) => a + b, 0);
    let a = input.map((row) => row.map((v) => 1 - v));
    for (const axis of [0, 1]) {
      const out = Array.from({ length: n }, () => Array(n).fill(0));
      for (let y = 0; y < n; y++)
        for (let x = 0; x < n; x++)
          for (let j = -r; j <= r; j++) {
            const yy = axis === 0 ? y + j : y,
              xx = axis === 1 ? x + j : x;
            if (xx >= 0 && xx < n && yy >= 0 && yy < n)
              out[y][x] += (a[yy][xx] * k[j + r]) / sum;
          }
      a = out;
    }
    return a.map((row) => row.map((v) => Math.max(0, Math.min(1, 1 - v))));
  },
  bin(a, b) {
    const n = a.length / b;
    return Array.from({ length: n }, (_, y) =>
      Array.from({ length: n }, (_, x) => {
        let sum = 0;
        for (let j = 0; j < b; j++)
          for (let i = 0; i < b; i++) sum += a[y * b + j][x * b + i];
        return sum / (b * b);
      }),
    );
  },
  solve(a, b) {
    const n = b.length,
      m = a.map((r, i) => [...r, b[i]]);
    for (let i = 0; i < n; i++) {
      let p = i;
      for (let j = i + 1; j < n; j++)
        if (Math.abs(m[j][i]) > Math.abs(m[p][i])) p = j;
      if (Math.abs(m[p][i]) < 1e-13) return null;
      [m[p], m[i]] = [m[i], m[p]];
      const d = m[i][i];
      for (let k = i; k <= n; k++) m[i][k] /= d;
      for (let j = 0; j < n; j++)
        if (j !== i) {
          const v = m[j][i];
          for (let k = i; k <= n; k++) m[j][k] -= v * m[i][k];
        }
    }
    return m.map((r) => r[n]);
  },
  fit(x, y, errors = null) {
    const no = (reason) => ({
      available: false,
      reason,
      phase: null,
      spacing: null,
      contrast: null,
    });
    if (x.length < 20 || y.some((v) => v === null || !Number.isFinite(v)))
      return no("Need at least 20 valid profile samples in the ROI.");
    const range = x.at(-1) - x[0],
      pitch = x[1] - x[0],
      max = Math.max(...y),
      min = Math.min(...y),
      scale = max - min;
    if (!(scale > 1e-10)) return no("No density modulation.");
    const positive = y.map((v) => Math.max(v, 0)),
      sum = positive.reduce((a, b) => a + b, 0);
    if (sum <= 0) return no("No positive signal.");
    const center = x.reduce((a, v, i) => a + v * positive[i], 0) / sum,
      sd = Math.sqrt(
        x.reduce((a, v, i) => a + (v - center) ** 2 * positive[i], 0) / sum,
      );
    const kmin = (2 * Math.PI * 2.5) / range,
      kmax = (2 * Math.PI) / (4 * pitch);
    if (kmax <= kmin || sd < pitch)
      return no("Insufficient spatial resolution.");
    const normalized = y.map((v) => v / scale);
    function trial(k, width, offset) {
      const rows = x.map((v) => {
        const z = (v - offset) / width,
          e = Math.exp(-0.5 * z * z);
        return [e, e * z, e * z * z, e * Math.cos(k * v), e * Math.sin(k * v)];
      });
      const a = Array.from({ length: 5 }, () => Array(5).fill(0)),
        b = Array(5).fill(0);
      for (let i = 0; i < x.length; i++)
        for (let j = 0; j < 5; j++) {
          b[j] += rows[i][j] * normalized[i];
          for (let q = 0; q < 5; q++) a[j][q] += rows[i][j] * rows[i][q];
        }
      const coeff = AbsorptionCamera3D.solve(a, b);
      if (!coeff) return null;
      const predicted = rows.map((r) =>
        r.reduce((v, t, i) => v + t * coeff[i], 0),
      );
      const loss = predicted.reduce(
        (v, t, i) => v + (t - normalized[i]) ** 2,
        0,
      );
      return { k, width, offset, coeff, predicted, loss, rows };
    }
    let best = null;
    const dk = (kmax - kmin) / 80;
    for (const width of [0.45, 0.6, 0.8, 1, 1.3, 1.7, 2.2].map((v) => v * sd))
      for (const offset of [center - 0.2 * sd, center, center + 0.2 * sd])
        for (let j = 0; j <= 80; j++) {
          const v = trial(kmin + j * dk, width, offset);
          if (v && (!best || v.loss < best.loss)) best = v;
        }
    if (!best) return no("Ill-conditioned profile fit.");
    for (let stage = 0; stage < 5; stage++) {
      const old = best,
        step = dk / 4 ** (stage + 1);
      for (const width of [1 - 0.15 / 2 ** stage, 1, 1 + 0.15 / 2 ** stage].map(
        (v) => v * old.width,
      ))
        for (const offset of [
          old.offset - pitch / 2 ** stage,
          old.offset,
          old.offset + pitch / 2 ** stage,
        ])
          for (let j = -5; j <= 5; j++) {
            const k = old.k + j * step;
            if (k <= kmin || k >= kmax) continue;
            const v = trial(k, width, offset);
            if (v && v.loss < best.loss) best = v;
          }
    }
    const amplitude = Math.hypot(best.coeff[3], best.coeff[4]),
      baseline = best.coeff[0],
      contrast = amplitude / baseline,
      rmse = Math.sqrt(best.loss / x.length);
    const observedError = errors ? Math.max(...errors) : 0;
    // Compare with the same smooth envelope with no carrier, fitted independently.
    const ids = [0, 1, 2],
      a = ids.map((i) =>
        ids.map((j) => best.rows.reduce((s, r) => s + r[i] * r[j], 0)),
      ),
      b = ids.map((i) =>
        best.rows.reduce((s, r, j) => s + r[i] * normalized[j], 0),
      );
    const coeff = this.solve(a, b),
      baseLoss = coeff
        ? best.rows.reduce(
            (s, r, j) =>
              s +
              (ids.reduce((v, id, k) => v + r[id] * coeff[k], 0) -
                normalized[j]) **
                2,
            0,
          )
        : 0;
    if (
      baseline <= 0 ||
      contrast < 0.08 ||
      contrast > 1.1 ||
      rmse > 0.12 ||
      baseLoss <= 0 ||
      1 - best.loss / baseLoss < 0.7 ||
      amplitude * scale < 5 * observedError ||
      best.k - kmin < dk / 4 ||
      kmax - best.k < dk / 4
    )
      return {
        ...no(
          "Fringes are unresolved, noisy, or incompatible with the envelope/carrier fit.",
        ),
        relative_rmse: rmse,
      };
    return {
      available: true,
      reason:
        "Image-only envelope/carrier fit; phase at horizontal coordinate zero.",
      phase: Math.atan2(best.coeff[4], best.coeff[3]),
      spacing: (2 * Math.PI) / best.k,
      contrast,
      relative_rmse: rmse,
      fit_profile: best.predicted.map((v) => v * scale),
    };
  },
  acquire(snapshot, settings) {
    const c = { ...this.defaults, ...settings },
      n = snapshot.config.n;
    this.validate(c, n);
    const axes = { z: ["x", "y", 0], y: ["x", "z", 1], x: ["y", "z", 2] }[
      c.axis
    ];
    const a = snapshot.scales.length_um,
      dx = (snapshot.config.length / n) * a,
      b = c.binning,
      pitch = dx * b;
    const fineX = snapshot.x.map((v) => v * a),
      x = Array.from(
        { length: n / b },
        (_, i) =>
          fineX.slice(i * b, (i + 1) * b).reduce((s, v) => s + v, 0) / b,
      );
    const select = x
        .map((v, i) => (Math.abs(v) <= c.roi_um ? i : -1))
        .filter((i) => i >= 0),
      strip = x
        .map((v, i) => (Math.abs(v) <= c.strip_um / 2 ? i : -1))
        .filter((i) => i >= 0);
    if (select.length < 4 || !strip.length)
      throw Error("Enlarge the ROI/strip or use smaller pixels.");
    const truth = snapshot.columns[axes[2]].map((r) =>
      r.map((v) => (v * snapshot.config.atoms) / a ** 2),
    );
    const t = truth.map((r) =>
      r.map((v) => this.transmission(this.sigma * v, c.saturation)),
    );
    const sampled = this.bin(
      this.blur(t, c.fwhm_um / (Math.sqrt(8 * Math.log(2)) * dx)),
      b,
    );
    const fluence =
      ((c.saturation * 16.69 * c.exposure_us * 1e-6) /
        ((6.62607015e-34 * 299792458) / 0.780241209e-6)) *
      1e-12;
    const reference = fluence * pitch * pitch * c.efficiency,
      rng = this.rng(c.seed),
      frame = (mean) =>
        100 +
        (c.noise ? rng.poisson(mean) + c.read_noise * rng.normal() : mean);
    const atoms = sampled.map((r) => r.map((v) => frame(v * reference))),
      ref = sampled.map((r) => r.map(() => frame(reference))),
      dark = sampled.map((r) =>
        r.map(() => 100 + (c.noise ? c.read_noise * rng.normal() : 0)),
      );
    const density = atoms.map((r, j) =>
      r.map((v, i) => {
        const A = v - dark[j][i],
          R = ref[j][i] - dark[j][i];
        return A > 0 && R > 0
          ? (Math.log(R / A) + (c.saturation * (R - A)) / reference) /
              this.sigma
          : null;
      }),
    );
    const variance = atoms.map((r, j) =>
      r.map((v, i) => {
        const A = v - dark[j][i],
          R = ref[j][i] - dark[j][i];
        if (density[j][i] === null) return null;
        if (!c.noise) return 0;
        const ga = -1 / A - c.saturation / reference,
          gr = 1 / R + c.saturation / reference,
          gd = 1 / A - 1 / R;
        return (
          (ga * ga * (A + c.read_noise ** 2) +
            gr * gr * (R + c.read_noise ** 2) +
            gd * gd * c.read_noise ** 2) /
          this.sigma ** 2
        );
      }),
    );
    const profile = x.map((_, i) =>
      strip.some((j) => density[j][i] === null)
        ? null
        : strip.reduce((s, j) => s + density[j][i], 0) / strip.length,
    );
    const err = x.map((_, i) =>
      strip.some((j) => variance[j][i] === null)
        ? null
        : Math.sqrt(strip.reduce((s, j) => s + variance[j][i], 0)) /
          strip.length,
    );
    const truthPixels = this.bin(truth, b),
      truthProfile = x.map(
        (_, i) =>
          strip.reduce((s, j) => s + truthPixels[j][i], 0) / strip.length,
      );
    const fitX = select.map((i) => x[i]),
      measuredFit = this.fit(
        fitX,
        select.map((i) => profile[i]),
        c.noise ? select.map((i) => err[i]) : null,
      ),
      truthFit = this.fit(
        fitX,
        select.map((i) => truthProfile[i]),
      );
    let count = 0,
      truthCount = 0,
      countVariance = 0,
      invalid = 0;
    for (const j of select)
      for (const i of select) {
        truthCount += truthPixels[j][i] * pitch ** 2;
        if (density[j][i] === null) invalid++;
        else {
          count += density[j][i] * pitch ** 2;
          countVariance += variance[j][i] * pitch ** 4;
        }
      }
    const total = truth.flat().reduce((s, v) => s + v, 0) * dx * dx;
    const absorbed =
      (fluence * t.flat().reduce((s, v) => s + 1 - v, 0) * dx * dx) /
      snapshot.config.atoms;
    const warnings = [];
    if (invalid)
      warnings.push(
        `${invalid} invalid ROI pixels; atom count and image fit may be unavailable. Magenta pixels are invalid; negative estimates remain blue.`,
      );
    if (!measuredFit.available) warnings.push(measuredFit.reason);
    if (c.axis === "x")
      warnings.push(
        "Looking along the split direction integrates x-fringes away; no x phase can be recovered.",
      );
    if (c.fwhm_um > 0 && c.fwhm_um < 2 * dx)
      warnings.push(
        "PSF is narrower than two source cells; refine the source grid for optical accuracy.",
      );
    if (Math.max(...truth.flat()) * this.sigma > 4)
      warnings.push(
        "Peak optical depth exceeds 4; dark pixels and omitted multiple scattering limit interpretation.",
      );
    if (absorbed > 5)
      warnings.push(
        "More than five absorbed photons per atom on average; recoil and motion are not simulated.",
      );
    return {
      camera: c,
      axes: axes.slice(0, 2),
      x_um: x,
      pixel_um: pitch,
      truth_density: truthPixels,
      density,
      atoms_frame: atoms,
      reference_frame: ref,
      dark_frame: dark,
      expected_reference_e: reference,
      profile,
      profile_error: err,
      truth_profile: truthProfile,
      fit_x_um: fitX,
      measured: measuredFit,
      truth: truthFit,
      atoms_roi: invalid ? null : count,
      truth_atoms_roi: truthCount,
      atom_standard_error: invalid ? null : Math.sqrt(countVariance),
      invalid_roi_pixels: invalid,
      model_total_atoms: total,
      scattered_per_atom: absorbed,
      roi_bounds_um: [x[select[0]] - pitch / 2, x[select.at(-1)] + pitch / 2],
      strip_bounds_um: [x[strip[0]] - pitch / 2, x[strip.at(-1)] + pitch / 2],
      warnings,
    };
  },
};

File: coldatomlab/camera3d.py
"""Independent NumPy camera oracle and browser acquisition verification.

Projection is N integral |psi|² dX_los / a0², never a 2D frozen-axis reduction.
"""

import math

import numpy as np

from .imaging import SIGMA_UM2, bin_mean, blur_transmission, transmission
from .solver3d import Config3D


class CameraRNG:
    """Specified 32-bit LCG, Box-Muller normals, Knuth/PTRS Poisson counts."""

    def __init__(self, seed):
        self.state = seed

    def uniform(self):
        self.state = (1664525 * self.state + 1013904223) & 0xFFFFFFFF
        return (self.state + 0.5) / 4294967296

    def normal(self):
        return math.sqrt(-2 * math.log(self.uniform())) * math.cos(2 * math.pi * self.uniform())

    def poisson(self, mu):
        if mu < 30:
            k, p, bound = 0, 1.0, math.exp(-mu)
            while True:
                k += 1
                p *= self.uniform()
                if p <= bound:
                    return k - 1
        b = 0.931 + 2.53 * math.sqrt(mu)
        a, inv, vr = -0.059 + 0.02483 * b, 1.1239 + 1.1328 / (b - 3.4), 0.9277 - 3.6224 / (b - 2)
        while True:
            u, v = self.uniform() - 0.5, self.uniform()
            us = 0.5 - abs(u)
            k = math.floor((2 * a / us + b) * u + mu + 0.43)
            if us >= 0.07 and v <= vr:
                return k
            if k < 0 or (us < 0.013 and v > us):
                continue
            logfact = (
                sum(math.log(j) for j in range(2, k + 1))
                if k < 256
                else (
                    (k + 0.5) * math.log(k)
                    - k
                    + 0.5 * math.log(2 * math.pi)
                    + 1 / (12 * k)
                    - 1 / (360 * k**3)
                    + 1 / (1260 * k**5)
                )
            )
            if math.log(v * inv / (a / (us * us) + b)) <= -mu + k * math.log(mu) - logfact:
                return k


def project(source, axis):
    c = Config3D(**source["config"])
    psi = np.asarray(source["psi_real"]) + 1j * np.asarray(source["psi_imag"])
    if psi.shape != (c.n,) * 3 or not np.isfinite(psi).all():
        raise ValueError("Invalid camera source field.")
    directions = {"x": (0, ["y", "z"]), "y": (1, ["x", "z"]), "z": (2, ["x", "y"])}
    if axis not in directions:
        raise ValueError("Invalid camera direction.")
    along, axes = directions[axis]
    density = (
        (abs(psi) ** 2).sum(axis=along).T * (c.length / c.n) * c.atoms / c.scales["length_um"] ** 2
    )
    return density, (np.arange(c.n) - c.n // 2) * (c.length / c.n) * c.scales["length_um"], axes


def fit_profile(x, y, errors=None):
    """Image-only scan, using independent NumPy least squares rather than JS elimination."""

    def no(reason):
        return dict(available=False, reason=reason, phase=None, spacing=None, contrast=None)

    x, y = np.asarray(x), np.asarray(y, dtype=float)
    if len(x) < 20 or not np.isfinite(y).all():
        return no("Need 20 valid samples.")
    span, pitch, scale = np.ptp(x), x[1] - x[0], np.ptp(y)
    if scale <= 1e-10 or np.maximum(y, 0).sum() <= 0:
        return no("No signal.")
    p = np.maximum(y, 0)
    center = np.dot(x, p) / sum(p)
    sd = math.sqrt(np.dot((x - center) ** 2, p) / sum(p))
    kmin, kmax = 2 * np.pi * 2.5 / span, 2 * np.pi / (4 * pitch)
    if kmax <= kmin or sd < pitch:
        return no("Insufficient resolution.")
    yn = y / scale

    def trial(k, width, offset):
        z = (x - offset) / width
        e = np.exp(-0.5 * z * z)
        matrix = np.array([e, e * z, e * z * z, e * np.cos(k * x), e * np.sin(k * x)]).T
        coeff, _, rank, _ = np.linalg.lstsq(matrix, yn, rcond=None)
        if rank < 5:
            return None
        predicted = matrix @ coeff
        return dict(
            k=k,
            width=width,
            offset=offset,
            coeff=coeff,
            predicted=predicted,
            loss=sum((predicted - yn) ** 2),
            rows=matrix,
        )

    best = None
    dk = (kmax - kmin) / 80
    for width in sd * np.array([0.45, 0.6, 0.8, 1, 1.3, 1.7, 2.2]):
        for offset in [center - 0.2 * sd, center, center + 0.2 * sd]:
            for j in range(81):
                value = trial(kmin + j * dk, width, offset)
                if value is not None and (best is None or value["loss"] < best["loss"]):
                    best = value
    if best is None:
        return no("Ill-conditioned fit.")
    for stage in range(5):
        old, step = best, dk / 4 ** (stage + 1)
        for width in old["width"] * np.array([1 - 0.15 / 2**stage, 1, 1 + 0.15 / 2**stage]):
            for offset in [
                old["offset"] - pitch / 2**stage,
                old["offset"],
                old["offset"] + pitch / 2**stage,
            ]:
                for j in range(-5, 6):
                    k = old["k"] + j * step
                    if not kmin < k < kmax:
                        continue
                    value = trial(k, width, offset)
                    if value is not None and value["loss"] < best["loss"]:
                        best = value
    amplitude = math.hypot(*best["coeff"][3:5])
    baseline = best["coeff"][0]
    contrast = amplitude / baseline if baseline else float("inf")
    rmse = math.sqrt(best["loss"] / len(x))
    smooth = best["rows"][:, [0, 1, 2]]
    coeff = np.linalg.lstsq(smooth, yn, rcond=None)[0]
    base_loss = sum((smooth @ coeff - yn) ** 2)
    noise = max(errors) if errors is not None else 0
    if (
        baseline <= 0
        or not 0.08 <= contrast <= 1.1
        or rmse > 0.12
        or base_loss <= 0
        or 1 - best["loss"] / base_loss < 0.7
        or amplitude * scale < 5 * noise
        or best["k"] - kmin < dk / 4
        or kmax - best["k"] < dk / 4
    ):
        return no("Unresolved/noisy/model mismatch.")
    return dict(
        available=True,
        phase=math.atan2(best["coeff"][4], best["coeff"][3]),
        spacing=2 * np.pi / best["k"],
        contrast=contrast,
        relative_rmse=rmse,
        fit_profile=(best["predicted"] * scale).tolist(),
    )


def acquire_reference(source, c):
    density, fine_x, axes = project(source, c["axis"])
    return acquire_projection(density, fine_x, axes, c, source["config"]["atoms"])


def validate_camera(c, n):
    """Validate browser-camera settings without acquiring a frame."""
    if c["axis"] not in ("x", "y", "z"):
        raise ValueError("Invalid camera direction.")
    b = c["binning"]
    if type(b) is not int or b not in (1, 2, 4, 8, 16) or n % b or n // b < 8:
        raise ValueError("Invalid binning.")
    for key, lo, hi in [
        ("fwhm_um", 0, 20),
        ("saturation", 0.001, 10),
        ("exposure_us", 0.1, 100),
        ("efficiency", 0.05, 1),
        ("read_noise", 0, 20),
        ("roi_um", 1, 100),
        ("strip_um", 0.2, 40),
    ]:
        if (
            isinstance(c[key], bool)
            or not isinstance(c[key], (int, float))
            or not math.isfinite(c[key])
            or not lo <= c[key] <= hi
        ):
            raise ValueError(f"Invalid {key}.")
    if (
        type(c["noise"]) is not bool
        or type(c["seed"]) is not int
        or not 0 <= c["seed"] <= 2**32 - 1
    ):
        raise ValueError("Invalid noise settings.")


def acquire_projection(density, fine_x, axes, c, atom_number, fit_fringes=True):
    """Independent optical acquisition from a verified physical column density."""
    b = c["binning"]
    n = len(fine_x)
    validate_camera(c, n)
    dx = fine_x[1] - fine_x[0]
    pitch = dx * b
    x = fine_x.reshape(-1, b).mean(axis=1)
    t = transmission(SIGMA_UM2 * density, c["saturation"])
    sampled = bin_mean(blur_transmission(t, c["fwhm_um"] / (math.sqrt(8 * math.log(2)) * dx)), b)
    fluence = (
        c["saturation"]
        * 16.69
        * c["exposure_us"]
        * 1e-6
        / (6.62607015e-34 * 299792458 / (0.780241209e-6))
        * 1e-12
    )
    reference = fluence * pitch**2 * c["efficiency"]
    rng = CameraRNG(c["seed"])

    def frame(means):
        return np.array(
            [
                100
                + (rng.poisson(float(mu)) + c["read_noise"] * rng.normal() if c["noise"] else mu)
                for mu in means.ravel()
            ]
        ).reshape(means.shape)

    atoms = frame(sampled * reference)
    ref = frame(np.full_like(sampled, reference))
    dark = np.array(
        [100 + (c["read_noise"] * rng.normal() if c["noise"] else 0) for _ in sampled.ravel()]
    ).reshape(sampled.shape)
    av, rv = atoms - dark, ref - dark
    valid = (av > 0) & (rv > 0)
    recovered = np.full_like(atoms, np.nan)
    recovered[valid] = (
        np.log(rv[valid] / av[valid]) + c["saturation"] * (rv[valid] - av[valid]) / reference
    ) / SIGMA_UM2
    variance = np.zeros_like(atoms)
    if c["noise"]:
        a, r = av[valid], rv[valid]
        ga, gr, gd = (
            -1 / a - c["saturation"] / reference,
            1 / r + c["saturation"] / reference,
            1 / a - 1 / r,
        )
        variance[valid] = (
            ga**2 * (a + c["read_noise"] ** 2)
            + gr**2 * (r + c["read_noise"] ** 2)
            + gd**2 * c["read_noise"] ** 2
        ) / SIGMA_UM2**2
    variance[~valid] = np.nan
    selection = np.abs(x) <= c["roi_um"]
    strip = np.abs(x) <= c["strip_um"] / 2
    if selection.sum() < 4 or not strip.any():
        raise ValueError("ROI/strip too small.")
    profile = recovered[strip].mean(axis=0)
    errors = np.sqrt(variance[strip].sum(axis=0)) / strip.sum()
    truth = bin_mean(density, b)
    truth_profile = truth[strip].mean(axis=0)
    roi = selection[:, None] & selection[None, :]
    invalid = int((~valid & roi).sum())
    return dict(
        axes=axes,
        x_um=x,
        truth_density=truth,
        density=recovered,
        **({} if fit_fringes else {"density_variance": variance}),
        atoms_frame=atoms,
        reference_frame=ref,
        dark_frame=dark,
        expected_reference_e=reference,
        pixel_um=pitch,
        model_total_atoms=density.sum() * dx**2,
        profile=profile,
        profile_error=errors,
        truth_profile=truth_profile,
        fit_x_um=x[selection],
        measured=(
            fit_profile(x[selection], profile[selection], errors[selection] if c["noise"] else None)
            if fit_fringes
            else dict(
                available=False,
                reason="Fringe analysis not requested.",
                phase=None,
                spacing=None,
                contrast=None,
            )
        ),
        truth=fit_profile(x[selection], truth_profile[selection])
        if fit_fringes
        else dict(
            available=False,
            reason="Fringe analysis not requested.",
            phase=None,
            spacing=None,
            contrast=None,
        ),
        atoms_roi=None if invalid else recovered[roi].sum() * pitch**2,
        truth_atoms_roi=truth[roi].sum() * pitch**2,
        atom_standard_error=None if invalid else math.sqrt(variance[roi].sum()) * pitch**2,
        invalid_roi_pixels=invalid,
        scattered_per_atom=fluence * (1 - t).sum() * dx**2 / atom_number,
        roi_bounds_um=[x[selection][0] - pitch / 2, x[selection][-1] + pitch / 2],
        strip_bounds_um=[x[strip][0] - pitch / 2, x[strip][-1] + pitch / 2],
    )


def verify_camera(data):
    if data.get("camera_model") != "rb87-browser-camera-v1":
        raise ValueError("Unknown 3D camera model.")
    saved = data["image"]
    regenerated = acquire_reference(data["source"], saved["camera"])
    return verify_image(saved, regenerated)


def verify_image(saved, regenerated):
    """Check numerical images and fits; reason/warning prose is not verified."""
    if saved["axes"] != regenerated["axes"]:
        raise ValueError("Camera plane axes do not match the projection.")
    for key, value in regenerated.items():
        if key in ("axes", "measured", "truth"):
            continue
        actual = np.asarray(saved[key], dtype=float)
        expected = np.asarray(value, dtype=float)
        if actual.shape != expected.shape or not np.allclose(
            actual, expected, rtol=2e-10, atol=1e-8, equal_nan=True
        ):
            raise ValueError(f"3D camera replay differs in {key}.")
    for key in ("measured", "truth"):
        fit = regenerated[key]
        if saved[key]["available"] != fit["available"]:
            raise ValueError("Camera fit availability differs from independent reference.")
        if fit["available"]:
            if not all(np.isfinite(saved[key][k]) for k in ("phase", "spacing", "contrast")):
                raise ValueError("Nonfinite camera fit.")
            phase = abs(np.angle(np.exp(1j * (saved[key]["phase"] - fit["phase"]))))
            if (
                phase > 0.01
                or abs(saved[key]["spacing"] / fit["spacing"] - 1) > 0.005
                or abs(saved[key]["contrast"] - fit["contrast"]) > 0.01
            ):
                raise ValueError("Camera fit differs from independent reference.")
            curve = np.asarray(saved[key]["fit_profile"], dtype=float)
            if curve.shape != np.shape(fit["fit_profile"]) or not np.allclose(
                curve, fit["fit_profile"], rtol=1e-4, atol=1e-6
            ):
                raise ValueError("Camera fit curve differs from independent reference.")
            if not np.isclose(saved[key]["relative_rmse"], fit["relative_rmse"], atol=1e-5):
                raise ValueError("Camera fit residual differs from independent reference.")
        elif any(saved[key][k] is not None for k in ("phase", "spacing", "contrast")):
            raise ValueError("Unavailable camera fit must have null measurements.")
    return {
        "camera_verified": True,
        "frame_tolerance": 1e-8,
        "independent_fits": {key: regenerated[key] for key in ("measured", "truth")},
    }
